src/exo1 - Basic types/exo1.effect.solution.ts

```diff
@@ -0,0 +1,117 @@
+    return Option.none();
+  }
+
+  return Option.some(a / b);
+};
+
+// You probably wrote `safeDivide` using `if` statements, and it's perfectly valid!
+// There are ways to not use `if` statements.
+// Keep in mind that extracting small functions out of pipes and using `if` statements in them
+// is perfectly fine and is sometimes more readable than not using `if`.
+//
+// BONUS: Try now to re-write `safeDivide` without any `if`
+//
+// HINT: Have a look at `fromPredicate` constructor
+
+export const safeDivideBonus = (a: number, b: number): Option.Option<number> => {
+  // First create an Option from b, ensuring it's not zero
+  const nonZeroB = pipe(
+    Option.some(b),
+    Option.filter((n: number) => n !== 0)
+  );
+  
+  // Then map it to perform the division
+  return pipe(
+    nonZeroB,
+    Option.map((b) => a / b)
+  );
+};
+
+///////////////////////////////////////////////////////////////////////////////
+//                                  EITHER                                   //
+///////////////////////////////////////////////////////////////////////////////
+
+// Write the safe version of `divide` with signature:
+// safeDivideWithError : (a: number, b: number) => Either<DivideByZeroError, number>
+//
+// BONUS POINT: Implement `safeDivideWithError` in terms of `safeDivide`.
+//
+// HINT : Either has two basic constructors:
+// - `E.left(leftValue)`
+// - `E.right(rightValue)`
+// as well as "smarter" constructors like:
+// - `E.fromOption(() => leftValue)(option)`
+
+// Here is a simple error type to help you:
+export type DivisionByZeroError = 'Error: Division by zero';
+export const DivisionByZero = 'Error: Division by zero' as const;
+
+// Note: In Effect, Either is Either<Right, Left> (opposite to fp-ts)
+export const safeDivideWithError = (a: number, b: number): Either.Either<number, DivisionByZeroError> => {
+  return pipe(
+    safeDivide(a, b),
+    Either.fromOption(() => DivisionByZero)
+  );
+};
+
+///////////////////////////////////////////////////////////////////////////////
+//                                   EFFECT                                  //
+///////////////////////////////////////////////////////////////////////////////
+
+// Now let's say we have a (pretend) API call that will perform the division for us
+// (throwing an error when the denominator is 0)
+export const asyncDivide = async (a: number, b: number) => {
+  await sleep(1000);
+
+  if (b === 0) {
+    throw new Error('BOOM!');
+  }
+
+  return a / b;
+};
+
+// Write the safe version of `asyncDivide` with signature:
+// asyncSafeDivideWithError : (a: number, b: number) => Effect.Effect<never, DivideByZeroError, number>
+//
+// HINT: Effect has a special constructor to transform a Promise<T> into
+// an Effect<never, Error, T>:
+// - `Effect.tryPromise({try: () => promise, catch: reason => leftValue})`
+
+// Note: In Effect, Effect.Effect<A, E, R> where:
+// - A is the success type
+// - E is the error type
+// - R is the environment type
```

Comment:
Here: https://effect.website/docs/getting-started/the-effect-type/ they mention "required dependencies" or "requirements" ? Might be clearer than "environment type"